src/tvheadend/HTSPVFS.cpp

```diff
@@ -150,6 +167,12 @@ long long HTSPVFS::Size ( void )
   return ret;
 }
 
+bool HTSPVFS::IsRealTimeStream ( void )
+{
+  /* Set IsRealTimeStream when within last 10MB of an in-progress recording */
+  return (m_eofOffset > 0 && m_eofOffset < 10485760);
```

Comment:
Wouldn't it be better to calculate with a percentage here instead of a fixed (magic) byte value?

Reply:
@ksooo

Sounds good. To be honest I was struggling to decide on what condition to trigger IsRealTimeStream.

So....what %played of an in-progress recording would you recommend for toggling IsRealTimeStream?

Reply:
I wouldn't use a percentage. This will make the real-time window much shorter on a new recording and really long on a large recording.

I think fixed is the best, unless we can check the remaining seconds or something.

Reply:
Right, percentage is not a better approach; best would be 10 seconds till EOF. No idea how to calculate that, though.

Reply:
I fully agree that time-based would be best; I originally tried to find ways to calculate this...However it seems in the context of pvr plugin there is no concept of time when dealing with recordings; just filesizes and offsets.

I also wondered whether the tipping point should be a factor of readchunksize; but dont know the VideoPlayer side of things well enough.

Guess the question is...when playing in non-real time mode then how far ahead does it need to be able to read, without delay, in order to avoid buffering. A size, number of seconds, number of readchunksizes?

If it is 10 seconds i think best we can do is agreed on a max filesize which can represent 10s.

I experimented with different sizes and the point it seemed to stop buffering after skipping beyond eof was somewhere between 3-4mb on a hd recording, and around 8mb on uhd (astra demo uhd channel). This is where the 10mb came from. In my tests I also found that simply setting IsRealTimeStream permanently for all recordings works just fine too. Not suggesting we do that but just trying to get things in context.

Reply:
Looking further at this we can find epoch of when recording started using m_playingRecording->GetFilesStart().

Based on this, current epoch, and current file size, we can work out based on avg bitrate, how many bytes represents 10secs.

Will update PR when done.....

Reply:
👍

Reply:
I would use different values for entering realtime and leaving realtime. You don't want the flag to toggle when playing pointer is on the boundary.

Reply:
I've adjusted the PR to toggle isRealTimeStream based on whether playback is within 10 seconds (rather than 10mb) of EOF.

It only calculates this as part of Seek operations and stores the bool, so if playing pointer happens to be on the boundary then there's no danger of constantly toggling the flag.

Reply:
> It only calculates this as part of Seek operations

imo this is incomplete. you can pause a recording that is close to live and after a while it should become non realtime

Reply:
@FernetMenta Good spot... clearly will need to re-work this.
I had tried to avoid these calculations after every read when playing an in-progress recording, but maybe that's unavoidable.

Reply:
Just working on an update to the PR to recalculate m_isRealTimeStream:

1) Following a seek
2) Following each read, but only when m_isRealTimeStream is currently TRUE

If you're in non-realtime mode, then I don't see any way to move into realtime (<10s of EOF), other than a seek (forward) operation. But if you're currently in real-time, then there is obviously the possibility of pause/play to go back to non-realtime (>10s of EOF) which I'd overlooked.

Also, this is the only way I can see us avoiding the danger of constantly toggling the flag when on the boundary during playback.

Reply:
I've adjusted the PR as above

Reply:
Having tested my previous change I spotted a floor in the logic of triggering the re-calculation on each Read().

During playback kodi (obviously) reads ahead of the play position (updating m_offset as it goes).

Therefore, doing the same calculations (based on m_offset) after each Read() we are actually calculating how far the end of the read buffer (rather than our playback position) is from the end of the file.

The calculation (based on m_offset) at the end of Seek() is valid because we haven't yet done the first Read() following the Seek; thus m_offset correctly reflects the current playback position.

To correctly handle Pause/Play I've therefore made some changes based on the PauseStream() API call; storing the time that the stream was paused (m_timePaused), and later using this when playback resumes to add a number of seconds to the stored offset (m_eofOffsetSecs); which was calculated during previous Seek().

If a Seek() is done whilst paused then the same calculations are performed, and m_timePaused is additionally updated to the current time. Thus when playback resumes then we can update m_eofOffsetSecs to the correct value.

I've added a couple of extra test cases above to cover "Pause->Play" and "Pause->Seek->Play".

I plan to run through all those original tests again later this week, but would value feedback in the meantime.